scintillometry/io/psrfits_io.py

```diff
@@ -0,0 +1,297 @@
+    file object and the PSRFITS SUBINT HDU.
+
+    Parameters
+    ----------
+    primary_hdu : PsrfitsPrimaryHDU
+        The psrfits main header object
+    subint_hdu : HDU object
+        The psrfits data HDU.
+
+    Notes
+    -----
+    Right now we are assuming the data rows are continuous in time and the
+    frequency are the same.
+    """
+
+    _properties = ('start_time', 'sample_rate', 'shape', 'samples_per_frame',
+                   'polarization', 'frequency')
+
+    def __new__(cls, primary_hdu, subint_hdu=None):
+        # Map Subint subclasses
+        mode = primary_hdu.obs_mode
+        try:
+            cls = subint_map[mode]
+        except KeyError:
+            raise ValueError("'{}' is not a valid mode.".format(mode))
+        return super(SubintHDUBase, cls).__new__(cls)
+
+    def __init__(self, primary_hdu, subint_hdu=None):
+        self.primary_hdu = primary_hdu
+        super(SubintHDUBase, self).__init__(header=subint_hdu.header,
+                                            data=subint_hdu.data)
+        self.verify()
+        self.offset = 0
+
+    def verify(self):
+        if self.header['EXTNAME'].strip() != "SUBINT":
+            raise ValueError("Input HDU is not a SUBINT type.")
+
+    @property
+    def mode(self):
+        return self.primary_hdu.obs_mode
+
+    @property
+    def start_time(self):
+        # NOTE should we get the start time for each raw, in case the time gaps
+        # in between the rows
+        file_start = self.primary_hdu.start_time
+        return file_start
+
+    @property
+    def nrow(self):
+        return int(self.header['NAXIS2'])
+
+    @property
+    def nchan(self):
+        return int(self.header['NCHAN'])
+
+    @property
+    def npol(self):
+        return int(self.header['NPOL'])
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def shape(self):
+        raw_shape = self.raw_shape
+        new_shape = namedtuple('shape', ['nsample', 'nbin', 'nchan', 'npol'])
+        result = new_shape(raw_shape.nrow * raw_shape.samples_per_frame,
+                           raw_shape.nbin, raw_shape.nchan, raw_shape.npol)
+        return result
+
+    @property
+    def raw_shape(self):
+        r_shape = namedtuple('shape', ['nrow', 'samples_per_frame', 'nbin',
+                                       'nchan', 'npol'])
+        result = r_shape(self.nrow, self.samples_per_frame, self.nbin,
+                         self.nchan, self.npol)
+        return result
+
+    @property
+    def polarization(self):
+        return self.header['POL_TYPE']
+
+    @property
+    def frequency(self):
+        if 'DAT_FREQ' in self.columns.names:
+            freqs = u.Quantity(self.data['DAT_FREQ'],
+                               u.MHz, copy=False)[0]
+        else:
+            freqs = getattr(self.primary_hdu, 'frequency', None)
+        if freqs is not None:
+            freqs_shape = self.shape._replace(npol=1, nbin=1)[1:]
```

Comment:
`freq_shape = (self.nchan, 1)`

Reply:
We also need the dimension of phase bins. so I think to keep the old way would be better.

Reply:
You don't need leading `1` for broadcasting to work, so the suggestion should be fine.